course/test/src/matrix3_TEST.cc

```diff
@@ -0,0 +1,173 @@
+// This file describes a challenge of a C++ L1 Padawan. The goal
+// of this unit test is to suggest an API and the abstractions
+// needed to implement an isometry.
+
+// Consider including other header files if needed.
+// Copyright 2020 <Jose Tomas Lorente>
+
+
+#include <cmath>
+#include <sstream>
+#include <string>
+
+#include "gtest/gtest.h"
+#include <isometry/isometry.hpp>
+
+namespace ekumen {
+namespace math {
+namespace test {
+namespace {
+
+  GTEST_TEST(Matrix3Test, Matrix3FullTests) {
+    const double kTolerance{1e-12};
+    Matrix3 m1{{1., 2., 3.}, {4., 5., 6.}, {7., 8., 9.}};
+    const Matrix3 m2{1., 2., 3., 4., 5., 6., 7., 8., 9.};
+    const Matrix3 m3{1., 2., 3., 4., 5., 6., 7., 8., 9.};
```

Comment:
Make the Vector3 class movable, and follow the rule of 5 both in Vector3 and in Matrix3.

For the sake of the exercise make `Vector3` something that moving from is worth it. I suggest instead of storing the elements as "double x_,y_,z_;" store them in a struct:
```
struct Elements {
  double x_,y_,z_;
};
```
then in `Vector3`, declare the storage as:
```
private:
   Elements *v_;
```
such that you allocate the space for `v_` when you construct with `new` and deallocate it with `delete` in the destructor (if needed, beware when you move from an object).

The rest of the changes should be trivial, `x_` will become `v_->x`, `y_` will become`v_->y` and `z_` will become `v_->z`, and your test coverage will protect you against messing up during refactoring.

Again, make sure you follow the rule of five in `Vector3`, and then do the same in `Matrix3`. If you do it right, you'll see  no differences... if you mess the move copy operations you may have memory access problems and premature death of the test.

Notice you should add test cases for your move operations , to make sure your move operations are getting selected when you think they are; there are no hard rules for how to test that, but in your case you can add a "bool moved()" method that returns `true` if the Vector3 object is the result of a move operation/constructor, and `false` in other cases. You can then test its value in the resulting object after a copy/move test.

Reply:
Keep this in mind: https://stackoverflow.com/questions/7027523/what-can-i-do-with-a-moved-from-object

Reply:
> The rest of the changes should be trivial, x_ will become v_->x, y_ will becomev_->y and z_ will become v_->z, and your test coverage will protect you against messing up during refactoring.

Im having some issues with this Gera, all the lines are giving me 
` error: 'v_' was not declared in this scope
     v_->x_ *= obj;` D:

Reply:
> Im having some issues with this Gera, all the lines are giving me
It builds for me, are you still having issues?

Reply:
@glpuga Wut, after pulling?